akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -711,78 +719,60 @@ private[akka] final case class MapAsync[In, Out](parallelism: Int, f: In ⇒ Fut
+      //FIXME Put Supervision.stoppingDecider as a SupervisionStrategy on DefaultAttributes.mapAsync?
+      lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+      var buffer: BufferImpl[Holder[Out]] = _
+      val futureCB =
+        getAsyncCallback[Holder[Out]](
+          _.elem match {
+            case Failure(e) if decider(e) == Supervision.Stop ⇒ failStage(e)
+            case _ ⇒ if (isAvailable(out)) pushOne()
+          })
 
-    var buffer: BufferImpl[Holder[Try[Out]]] = _
-    def todo = buffer.used
+      private[this] def todo = buffer.used
 
-    override def preStart(): Unit = buffer = BufferImpl(parallelism, materializer)
+      override def preStart(): Unit = buffer = BufferImpl(parallelism, materializer)
 
-    @tailrec private def pushOne(): Unit =
-      if (buffer.isEmpty) {
-        if (isClosed(in)) completeStage()
-        else if (!hasBeenPulled(in)) pull(in)
-      } else if (buffer.peek.elem == NotYetThere) {
-        if (todo < parallelism && !hasBeenPulled(in)) tryPull(in)
-      } else buffer.dequeue().elem match {
-        case Failure(ex) ⇒ pushOne()
-        case Success(elem) ⇒
-          push(out, elem)
+      @tailrec private def pushOne(): Unit =
+        if (buffer.isEmpty) {
+          if (isClosed(in)) completeStage()
+          else if (!hasBeenPulled(in)) pull(in)
+        } else if (buffer.peek.elem == NotYetThere) {
           if (todo < parallelism && !hasBeenPulled(in)) tryPull(in)
-      }
-
-    def failOrPull(holder: Holder[Try[Out]], f: Failure[Out]) =
-      if (decider(f.exception) == Supervision.Stop) failStage(f.exception)
-      else {
-        holder.elem = f
-        if (isAvailable(out)) pushOne()
-      }
-
-    val futureCB =
-      getAsyncCallback[(Holder[Try[Out]], Try[Out])]({
-        case (holder, f: Failure[_]) ⇒ failOrPull(holder, f)
-        case (holder, s @ Success(elem)) ⇒
-          if (elem == null) {
-            val ex = ReactiveStreamsCompliance.elementMustNotBeNullException
-            failOrPull(holder, Failure(ex))
-          } else {
-            holder.elem = s
-            if (isAvailable(out)) pushOne()
-          }
-      })
+        } else buffer.dequeue().elem match {
+          case Success(elem) ⇒
+            push(out, elem)
+            if (todo < parallelism && !hasBeenPulled(in)) tryPull(in)
+          case Failure(ex) ⇒ pushOne()
+        }
 
-    setHandler(in, new InHandler {
       override def onPush(): Unit = {
         try {
           val future = f(grab(in))
-          val holder = new Holder[Try[Out]](NotYetThere)
+          val holder = new Holder[Out](NotYetThere, futureCB)
           buffer.enqueue(holder)
-          future.onComplete(result ⇒ futureCB.invoke(holder -> result))(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+          future.onComplete(holder)(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
```

Comment:
I'd still introduce `.fast` here until we get the improvements in scala... there's a ticket, we can discuss there :)